workspace/src/localization/localization_py/localization_py/EKF.py

```diff
@@ -0,0 +1,134 @@
+import numpy as np
+import math
+import yaml
+from dynamics import Dynamics
+
+
+class EKF:
+    """A basic Extended Kalman Filter implementation.
+    
+    This currently uses the 4DOF motion model. The Motion Model inputs are throttle and steering, and the observation model uses GPS and Magnetometer as sensors. States are cartesian coordinates x, y, heading theta, and velocity v.
+
+    Attributes:
+        Q: The covariane matrix for our states.
+        R: The covariance matrix for our sensors.
+        P: The predicted covariance estimate.
+        dt: The timestep for the filter.
+    """
+
+    def __init__(self, dt, dynamics, Q, R):
+        """Initialize the EKF.
+
+        Initialize each of the global variables for the EKF. Load the dynamics parameters for the vehicle from an external YAML file.
+
+        Args:
+            dt: The timestep at which this filter will operate at. 
+            dynamics_path: The path to the dynamics parameters yaml file.
+            param_path: The path to the filter parameters yaml file
+        """
+
+        self.dt = dt
+        self.dyn = Dynamics(dt, dynamics)
```

Comment:
Aaaaaaa I have some issues with this. The functional problem is that you are marrying the dt of the EKF with the dt of the dynamics model. This is almost never true (except for our case). The dynamics might need to run at a very high frequency, but because of low frequency of sensor feed, you might be able to do correction only once a while.

I think this can be solved in two ways, one of which I prefer much more than the other. 
1) In the "dynamics" list, also include dt. This way, you can have the EKF with a init that takes its own EKF dt and the dynamics list having its own dt. 
2) (the one I like) Rather than instantiating the class inside the EKF class, I think we should pass an instantiated dynamics class. Then, when you instantiate the dynamics class outside, you can specify a time step for it and pass the Dynamics instance to instantiate the EKF class. 

How will this change the downstream functions?
I think this should not be dealt with in the EKF class. Rather, I think the "motion_model" function should take an additional parameter that specifies the frequency at which it needs to provide a state. 

Let me know if this was unclear

Reply:
Hmm I'm not sure I understand the second option, as the EKF still needs to know how often to call the dynamics, so you anyways have to pass the "dynamics dt" to the EKF. Also, you can't propagate the dynamics outside of the filter as there are other matrices that the filter needs to maintain related to how often you call the dynamics.

For now, the control node doesn't provide inputs more often than 10hz; IMO we should merge this, and then open an issue about this and fix it in a seperate branch. This MR is already very fat...

Reply:
Yes lets open an issue on this later. You can pass the dynamics object directly which already bakes the information of the dynamics dt. This would be very useful in many situations - for example
The user already has a dynamics model that she can rely on. This dynamics model has the required functions to interact with the state estimation class. However, the dynamics class does not take an argument "dt". The way we have done it now, users are forced to use the dynamics model we provide because it has to take an argument dt. If the state estimation node directly takes a dynamics object, that object can be instantiated in any way the user likes - for all we care it could be a c++ wrapped model which reads a json file for parameters - as long as we can call the 2 required functions and we can qurey it for the timestep, we are good. Note that also dynamics models might not step at a single time step (variable time step integrators) but still provide states at the timesteps that the control/state estimation requires. In this case, dt is meaningless.